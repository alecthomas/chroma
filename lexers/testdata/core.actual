@pub use collections.{Array, BitSet, BitVec, HashMap, HashSet, List, Queue};

@pub mod annotations;

@pub @internal fun unreachable[T](): T

@pub fun unimplemented(): Unit = fatalError("not yet implemented")

fun isValidUtf8(data: Array[UInt8]): Bool {
  var i = 0i64

  while i < data.size {
    let by = data(i).toInt32
    var codepoint = 0i32
    var bytes = 0i32
    var min = 0i32
    var max = 0i32

    if by & 0x80i32 == 0i32 {
      bytes = 0
      i = i + 1i64
    } else {
      if by
      ... & 0xC0i32 == 0x80i32 {
        return false
      }
      ... & 0xE0i32 == 0xC0i32 {
        codepoint = by & 0x1Fi32
        bytes = 1
        min = 0x80
        max = 0x7FFi32
      }
      ... & 0xF0i32 == 0xE0i32 {
        codepoint = by & 0xFi32
        bytes = 2i32
        min = 0x800i32
        max = 0xFFFFi32
      }
      ... & 0xF8i32 == 0xF0i32 {
        codepoint = by & 0x7i32
        bytes = 3
        min = 0x10000
        max = 0x10FFFFi32
      }

      while bytes > 0i32 {
        bytes = bytes - 1i32
        i = i + 1i64

        if i >= data.size {
          return false
        }

        let by = data(i).toInt32

        if by & 0xC0i32 != 0x80i32 {
          return false
        }

        codepoint = codepoint.shiftLeft(6i32) | (by & 0x3Fi32)
      }

      if codepoint < min || codepoint > max {
        return false
      }

      i = i + 1i64
    }
  }

  return true
}

@internal fun unsafeKillRefs[T](arr: Array[T], idx: Int64): Unit

@pub class Stacktrace {
  backtrace: Option[Array[Int32]],
  elements: Option[Array[StacktraceElement]],
}

impl Stacktrace {
  @pub @static fun new(): Stacktrace {
    let trace = Stacktrace(None[Array[Int32]], None[Array[StacktraceElement]])
    trace.retrieveStacktrace()
    trace
  }

  @pub fun getStacktrace(): Array[StacktraceElement] {
    if self.elements.isSome {
      return self.elements.getOrPanic()
    }

    if self.backtrace.isNone {
      let elements = Array[StacktraceElement]::new()
      self.elements = Some[Array[StacktraceElement]](elements)
      return elements
    }

    let backtraceLength = self.backtrace.getOrPanic().size
    var i = 0i64
    let len = backtraceLength / 2i64
    let elements = List[StacktraceElement]::new()
    elements.reserve(len)

    while i < len {
      elements.push(self.getStacktraceElement(i))
      i = i + 1i64
    }

    let elements = elements.toArray()
    self.elements = Some(elements)
    return elements
  }

  @pub fun printStacktrace(): Unit {
    let x = self.getStacktrace()
    var i = 0i64

    for frame in self.getStacktrace() {
      println(frame.toString())
    }
  }

  @internal fun retrieveStacktrace(): Unit
  @internal fun getStacktraceElement(idx: Int64): StacktraceElement
}

class StacktraceElement {
  name: String,
  line: Int32,
}

impl StacktraceElement {
  fun toString(): String = "${self.name}: ${self.line}"
}

/* total order */
@pub trait Sortable {
  fun sortsAs(other: Self): Int32
  // ideally we would be able to provide a default implementation for the next 3 methods based on `sortsAs`
  fun sortsBefore(other: Self): Bool
  fun sortsAfter(other: Self): Bool
  fun sortsSame(other: Self): Bool
}

@pub enum Option[T] {
  Some(T),
  None,
}

use Option.Some
use Option.None

impl[T] Option[T] {
  @pub fun isNone: Bool = if self
    ... is Some(_) { false }
    ... is None    { true  }

  @pub fun isSome: Bool = if self
    ... is Some(_) { true  }
    ... is None    { false }
}

impl[T] Option[Option[T]] {
  // WTF -- removing `()` causes memory corruption in main (tests/stdlib/option-flatten.core:6) -- WTF
  @pub fun flatten(): Option[T] = if self
    ... is Some(val) { val }
    ... is None      { None[T] }
}

@pub enum Result[T, E] {
  Pass(T),
  Fail(E),
}

use Result.Pass
use Result.Fail

impl[T, E] Result[T, E] {
  @pub fun isPass: Bool = if self
    ... is Pass(_) { true  }
    ... is Fail(_) { false }

  @pub fun isFail: Bool = if self
    ... is Pass(_) { false }
    ... is Fail(_) { true  }
}
