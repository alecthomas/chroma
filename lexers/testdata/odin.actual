package main

import "core:fmt"

/*
some comments in odin can
 /* be nested! */
and it should still work
*/

Data :: struct {
	c: complex64,
	q: quaternion256,
}

/* some other comment */

E :: enum(u32) {
	First,
	Second,
	Third,
}

E_Set :: distinct bit_set[E; u32]

foo_int :: proc(x: int) -> bool {
	return false
}

foo_float :: proc(f: f32) -> bool {
	return true
}

foo_en :: proc(e: E) -> bool {
	return e == .Third
}

foo :: proc{foo_int, foo_float, foo_en}

SOME_CONSTANT :: #config(COMMAND_LINE_ARG, 3)
main :: proc() {
	loc := #caller_location

	m: map[string]struct{}
	da: [dynamic]f64

	len(da)
	cap(da)

	foo(32)

	#panic("oof")

	y := &da
	y^ = make([dynamic]f64, 100)
	defer delete(da)

	x := []int{1, 2, 3, 4}
	for v, i in x {
		fmt.println(i, "-", v)
	}

	z := E_Set{.First, .Second}
	z2 := E_Set{.Third}

	assert(z | z2 == {.First, .Second, .Third})
}

@(test)
a_test_proc :: proc(t: ^testing.T) {
	value: int
	value = 3
	testing.errnof("a format: %s", value)
}

@(disable = LOG_LEVEL >= .Debug)
debug_thing :: proc(x, y, z: int) {
	fmt.println(x, y, z)
}

@private
program := `
foo :: proc() {
	fmt.println("hellope!")
}
`

@(private = "file")
pkg: int

