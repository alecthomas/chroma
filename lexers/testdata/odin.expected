[
  {"type":"KeywordNamespace","value":"package"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"main"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"KeywordNamespace","value":"import"},
  {"type":"TextWhitespace","value":" "},
  {"type":"LiteralString","value":"\"core:fmt\""},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"CommentMultiline","value":"/*\nsome comments in odin can\n /* be nested! */\nand it should still work\n*/"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"Data"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"struct"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"c"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"complex64"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"q"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"quaternion256"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"CommentMultiline","value":"/* some other comment */"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"E"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"enum"},
  {"type":"Punctuation","value":"("},
  {"type":"KeywordType","value":"u32"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"First"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"Second"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"Third"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"E_Set"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Keyword","value":"distinct"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Keyword","value":"bit_set"},
  {"type":"Punctuation","value":"["},
  {"type":"Name","value":"E"},
  {"type":"Punctuation","value":";"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"u32"},
  {"type":"Punctuation","value":"]"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"foo_int"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"x"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"-\u003e"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"bool"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Keyword","value":"return"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordConstant","value":"false"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"foo_float"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"f"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"f32"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"-\u003e"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"bool"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Keyword","value":"return"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordConstant","value":"true"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"foo_en"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"e"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"E"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"-\u003e"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"bool"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Keyword","value":"return"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"e"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"=="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"Third"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"foo"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"{"},
  {"type":"Name","value":"foo_int"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"foo_float"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"foo_en"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"Name","value":"SOME_CONSTANT"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameDecorator","value":"#config"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"COMMAND_LINE_ARG"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameOther","value":"3"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Name","value":"main"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"()"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"loc"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":":="},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameDecorator","value":"#caller_location"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Name","value":"m"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"map"},
  {"type":"Punctuation","value":"["},
  {"type":"KeywordType","value":"string"},
  {"type":"Punctuation","value":"]"},
  {"type":"KeywordDeclaration","value":"struct"},
  {"type":"Punctuation","value":"{}"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"da"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"["},
  {"type":"Keyword","value":"dynamic"},
  {"type":"Punctuation","value":"]"},
  {"type":"KeywordType","value":"f64"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Keyword","value":"len"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"da"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Keyword","value":"cap"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"da"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Name","value":"foo"},
  {"type":"Punctuation","value":"("},
  {"type":"NameOther","value":"32"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"NameDecorator","value":"#panic"},
  {"type":"Punctuation","value":"("},
  {"type":"LiteralString","value":"\"oof\""},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Name","value":"y"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":":="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"\u0026"},
  {"type":"Name","value":"da"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"y"},
  {"type":"Operator","value":"^"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Keyword","value":"make"},
  {"type":"Punctuation","value":"(["},
  {"type":"Keyword","value":"dynamic"},
  {"type":"Punctuation","value":"]"},
  {"type":"KeywordType","value":"f64"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameOther","value":"100"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Keyword","value":"defer"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Keyword","value":"delete"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"da"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Name","value":"x"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":":="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"[]"},
  {"type":"KeywordType","value":"int"},
  {"type":"Punctuation","value":"{"},
  {"type":"NameOther","value":"1"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameOther","value":"2"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameOther","value":"3"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameOther","value":"4"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Keyword","value":"for"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"v"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"i"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Keyword","value":"in"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"x"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t\t"},
  {"type":"Name","value":"fmt"},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"println"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"i"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"LiteralString","value":"\"-\""},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"v"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Name","value":"z"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":":="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"E_Set"},
  {"type":"Punctuation","value":"{."},
  {"type":"Name","value":"First"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"Second"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"z2"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":":="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"E_Set"},
  {"type":"Punctuation","value":"{."},
  {"type":"Name","value":"Third"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n\t"},
  {"type":"Name","value":"assert"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"z"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"|"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"z2"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"=="},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{."},
  {"type":"Name","value":"First"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"Second"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"Third"},
  {"type":"Punctuation","value":"})"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"NameAttribute","value":"@(test)"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Name","value":"a_test_proc"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"t"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"^"},
  {"type":"Name","value":"testing"},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"T"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"value"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"value"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"="},
  {"type":"TextWhitespace","value":" "},
  {"type":"NameOther","value":"3"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"testing"},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"errnof"},
  {"type":"Punctuation","value":"("},
  {"type":"LiteralString","value":"\"a format: %s\""},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"value"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"NameAttribute","value":"@(disable = ODIN_DEBUG)"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Name","value":"debug_thing"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":"::"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordDeclaration","value":"proc"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"x"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"y"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"z"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Punctuation","value":"{"},
  {"type":"TextWhitespace","value":"\n\t"},
  {"type":"Name","value":"fmt"},
  {"type":"Punctuation","value":"."},
  {"type":"Name","value":"println"},
  {"type":"Punctuation","value":"("},
  {"type":"Name","value":"x"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"y"},
  {"type":"Punctuation","value":","},
  {"type":"TextWhitespace","value":" "},
  {"type":"Name","value":"z"},
  {"type":"Punctuation","value":")"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Punctuation","value":"}"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"NameAttribute","value":"@private\n"},
  {"type":"Name","value":"program"},
  {"type":"TextWhitespace","value":" "},
  {"type":"Operator","value":":="},
  {"type":"TextWhitespace","value":" "},
  {"type":"LiteralString","value":"`\nfoo :: proc() {\n\tfmt.println(\"hellope!\")\n}\n`"},
  {"type":"TextWhitespace","value":"\n\n"},
  {"type":"NameAttribute","value":"@(private = \"file\")"},
  {"type":"TextWhitespace","value":"\n"},
  {"type":"Name","value":"pkg"},
  {"type":"Operator","value":":"},
  {"type":"TextWhitespace","value":" "},
  {"type":"KeywordType","value":"int"},
  {"type":"TextWhitespace","value":"\n\n"}
]
